### PR TITLE
I7757 context help

### LINKED ISSUE #7757: Context Sensitive Help in NVDA Dialogs
### Steps to reproduce:

* NVDA has a well thought out user guide, but it is prohibitively long for new users. A new user needs to be able to get help relevant to what they are doing while they are working in various dialogs and controls with NVDA.
* Some controls in NVDA preferences dialogs have long labels which cause them not to fit well visually (see #6345). These labels can be shortened if additional help for these controls is readily available with F1 or by pressing a Help button.

### Expected behavior:

1. Explicit anchors should be added to all sections of the NVDA User Guide so that it can be opened to the correct place when context help is invoked.
2. When in an NVDA preference dialog a Help button should be available to open the NVDA User Guide  for all controls in that dialog.
3. Optionally, pressing F1 on any control within an NVDA Preference dialog should open the NVDA User Guide to the section describing the function of that specific control if one exists, or for the dialog if no explicit help exists.

### Actual behavior:

Currently NVDA's preferences dialogs do not have a help button and do not respond to the F1 key. The NVDA User Guide explains NVDA's dialogs and controls in detail, but new users find it daunting to locate information quickly.

### Similar issues which this is not:

* Providing context sensitive help for NVDA's dialogs is not a tutor mode as in #3232.
* This issue is not intended to provide general Windows help as in #2699.

### LINKED ISSUE #8354: Context Sensitive Help for NVDA
### Steps to reproduce:
NVDA has an extensive user guide, but it can be daunting to new users to locate the help for a particular control or dialog box in a monolythic help file.

### Expected behavior:
NVDA dialogs should include a help button as well as provide context sensitive help in response to a press of F1.

### Actual behavior:
Users must locate the appropriate section in the user guide which deals with the control or dialog they have questions about.

### System configuration:
NVDA version:
Suggestion for future release.

NVDA Installed or portable:
Both

Other information:
I will submit a pull request referencing this issue shortly.

Windows version:
Any

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
Fixes #7757
Fixes #8354
Supersedes PR #8355

### Summary of the issue:
Add context sensitive help referencing the user guide to settings dialogs, panels, and controls.

### Description of how this pull request fixes the issue:
This PR adds initial support for opening the user guide at a particular section based on the context when the user presses F1 (in NVDA dialogs). For the sake of simplicity, this is done by opening the user guide in the browser. Unfortunately, this results in a lot of unnecessary verbosity. However, this is still quicker than opening the user guide separately (also in a browser) and navigating to the chosen section manually. Further, this aspect of the help delivery UX can be further optimized later.

In #7757 a help button was requested, there are several UI design concerns around this. Instead of dwelling on this, I'd like to proceed with the feature.

### Testing performed:
- Tested that the correct section of help was shown in several dialogs. My default browser (Firefox) was opened, and NVDA could read that section.
- Tested that several controls with specific help also worked.

I did not test all of the controls. Hopefully alpha testers could focus on this a little bit to ensure there are no problems.

### Known issues with pull request:
The delivery of the help is not as smooth as it could potentially be. However, this could be improved with subsequent work.

### Change log entry:
New features:
`- Pressing F1 inside NVDA dialogs will now open the help file to most relevant section. (#7757)`

### CODE REVIEW

**Comment by feerrenrut** [comment]
Some suggestions for how to improve this in subsequent work:
- Use javascript in the userguide to hide unrelated sections.
- Use the wx HTML 2 (webview) control to embed the help file. I tried this (see commit 7542dcd), however getting NVDA into the webview was a problem. If I clicked on the control NVDA would load it and read it, but I gave up trying to make that happen automatically.

**Comment by lukaszgo1** [comment]
I'm afrait this does not work for me at all. Each time I press F1 the error sound plays and the following is in the log:

IO - inputCore.InputManager.executeGesture (22:30:40.585) - winInputHook (3432):
Input: kb(desktop):f1
DEBUG - gui.contextHelp.showHelp (22:30:40.585) - MainThread (7024):
Opening help: helpId = SpeechViewer
ERROR - unhandled exception (22:30:40.585) - MainThread (7024):
Traceback (most recent call last):
  File "gui\contextHelp.py", line 54, in <lambda>
    lambda evt: _onEvtHelp(helpId, evt),
  File "gui\contextHelp.py", line 61, in _onEvtHelp
    showHelp(helpId)
  File "gui\contextHelp.py", line 37, in showHelp
    contextHelpRedirect = os.path.join(os.path.dirname(helpFile), "contextHelp.html")
  File "C:\Python37\lib\ntpath.py", line 221, in dirname
    return split(p)[0]
  File "C:\Python37\lib\ntpath.py", line 183, in split
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType

**Comment by XLTechie** [comment]
My first thought was to use a CHM file version of the manual for this, and 
invoke the Windows help system to display it.

But, that aside, could we do something like the following?:

A script could be used at a pre-compile stage, to create a dict of line 
numbers, indicating the start and end of each section
That would be trivial to automate with any of the normal text processing tools. 
(sed, awk, etc., or Python itself)

When help is requested for a particular section, a method could look up the line 
number start-end entry for that section.

Then use something like the linecache module to extract that section from the 
manual on disk, and feed it to ui.browseableMessage in its HTML mode.

**Comment by dingpengyu** [comment]
> My first thought was to use a CHM file version of the manual for this, and invoke the Windows help system to display it. But, that aside, could we do something like the following?: A script could be used at a pre-compile stage, to create a dict of line numbers, indicating the start and end of each section That would be trivial to automate with any of the normal text processing tools. (sed, awk, etc., or Python itself) When help is requested for a particular section, a method could look up the line number start-end entry for that section. Then use something like the linecache module to extract that section from the manual on disk, and feed it to ui.browseableMessage in its HTML mode.
You can use sphinx to generate a dimension .chm file for the user guide.

**Comment by feerrenrut** [comment]
@lukaszgo1 

Hmm that is strange, it looks like it doesn't get a path for the user guide. Can you open the user guide from the tools menu?

**Comment by feerrenrut** [comment]
@lukaszgo1 I have just pushed a change that will help with debugging the issue you are seeing (I can't reproduce it). It adds logging of the path to the help file.

**Comment by lukaszgo1** [comment]
@feerrenrut I'm sorry for the false alarm I've tried this from sources but have forgotten to build user documentation first. Would it be possible to give a message in such case that the user guide has not been found?

File: user_docs/en/changes.t2t [hunk 1]
```diff
@@ -692,6 +692,7 @@ Highlights of this release include  support for charts in Microsoft word and Pow
 == Changes ==
 - The HIMS Braille Sense/Braille EDGE/Smart Beetle and Hims Sync Braille display drivers have been replaced by one driver. The new driver will automatically be activated for former syncBraille driver users. (#7459) 
  - Some keys , notably scroll keys, have been reassigned to follow the conventions used by Hims products. Consult the user guide for more details.
+- F1 can now be used in most NVDA dialogs to get context sensitive help.
```
Comment (lukaszgo1) [inline user_docs/en/changes.t2t:695]: I  believe Change log entry is added just before merge most of the time. It should also be in the section for NVDA 2020.3 or whatever is the most recent version at the time of merge, not in the section for 2018.3 as it is now.

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]

File: source/gui/contextHelp.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def _onEvtHelp(helpId: str, evt: wx.HelpEvent):
+	# Don't call evt.skip. Whe want more specific bindings to override less specific.
+	showHelp(helpId)
+
+
+class ContextHelpMixin:
+	def __init__(self, *args, **kwargs):
+		log.debug("reef")
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/contextHelp.py:66]: Haha yep, good catch.

File: source/gui/contextHelp.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	recognized control.
+	"""
+	if not helpId:
+		# Translators: Message indicating no context sensitive help is available.
+		noHelpMessage = _("No context sensitive help is available here at this time.")
+		ui.message(noHelpMessage)
+	helpFile = gui.getDocFilePath("userGuide.html")
+	log.debug(f"Opening help: helpId = {helpId}, userGuidePath: {helpFile}")
+
+	contextHelpRedirect = os.path.join(os.path.dirname(helpFile), "contextHelp.html")
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/gui/contextHelp.py:37]: Ah, yes. I'll move the file to the temp directory.

**Comment by feerrenrut** [comment]
This issues should all be resolved now. I think due to the redirect, NVDA does not always set focus to the anchor heading. This is probably something that should be fixed in NVDA. Pressing F6 (to focus the URL) then enter is a work-around.

I'd still like to get this PR merged. Since so many classes are touched this change set is quite prone to conflicts. The exact mechanism for displaying the help can be improved subsequently.

File: site_scons/site_tools/t2t.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
 
 def txt2tags_actionFunc(target,source,env):
 	import txt2tags
-	txt2tags.exec_command_line([str(source[0])])
+	txt2tags.exec_command_line(["--outfile", str(source[0])[:-3] + "html", str(source[0])])
```
Comment (leonardder) [inline site_scons/site_tools/t2t.py:17]: Why this change?

File: source/gui/addonGui.py [hunk 5]
```diff
@@ -145,6 +149,8 @@ def _instance(cls):
 		"""
 		return None
 
+	helpIds = "AddonsManager"
```
Comment (leonardder) [inline source/gui/addonGui.py:152]: Should probably be

```suggestion
	helpId = "AddonsManager"
```

File: source/gui/configProfiles.py [hunk 6]
```diff
@@ -380,11 +392,16 @@ def onClose(self, evt):
 		self.Parent.Enable()
 		self.Destroy()
 
-class NewProfileDialog(wx.Dialog):
+
+class NewProfileDialog(
+		gui.ContextHelpMixin,
+		wx.Dialog   # wxPython does not seem to call base class initializer, put last in MRO
+):
+	helpIds = "ConfigurationProfiles"
```
Comment (leonardder) [inline source/gui/configProfiles.py:400]:
```suggestion
	helpId = "ConfigurationProfiles"
```

File: source/gui/settingsDialogs.py [hunk 7]
```diff
@@ -268,15 +280,15 @@ class SettingsPanel(wx.Panel, DpiScalingHelperMixin, metaclass=guiHelper.SIPABCM
 
 	title=""
 	panelDescription=u""
+	helpId = ""
```
Comment (leonardder) [inline source/gui/settingsDialogs.py:283]: Should we leave this out?

File: source/gui/contextHelp.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	log.debug(f"Did context help binding for {window.__class__.__qualname__}")
+
+
+def _onEvtHelp(helpId: str, evt: wx.HelpEvent):
+	# Don't call evt.skip. Events bubble upwards through parent controls.
+	# Context help for more specific controls should override the less specific parent controls.
+	showHelp(helpId)
+
+
+class ContextHelpMixin:
```
Comment (leonardder) [inline source/gui/contextHelp.py:76]: Could you add the SIPABCMeta metaclass here?

File: source/gui/contextHelp.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	def __init__(self, *args, **kwargs):
+		super().__init__(*args, **kwargs)
+		helpId = getattr(self, "helpId", None)
```
Comment (leonardder) [inline source/gui/contextHelp.py:79]: I think helpId could be implemented as an abstract property. I also think every panel needs the mixin, because add-ons can't provide context help.
In any case, we should avoid that something that doesn't have help has the mixin.

**Review by LeonarddeR — DISMISSED** [review]

File: source/gui/settingsDialogs.py (hunk 7 above)
Comment (feerrenrut) [reply source/gui/settingsDialogs.py:283]: I think I should move it into the `ContextHelpMixin`.

File: source/gui/configProfiles.py (hunk 6 above)
Comment (feerrenrut) [reply source/gui/configProfiles.py:400]: Yep, good catch.

File: site_scons/site_tools/t2t.py (hunk 4 above)
Comment (feerrenrut) [reply site_scons/site_tools/t2t.py:17]: This was already on the branch, honestly I don't know why the change was made. It seems to do the same as the default (no args) command. I'll remove the change.

File: source/gui/contextHelp.py (hunk 8 above)
Comment (feerrenrut) [reply source/gui/contextHelp.py:76]: This is only necessary if we make the mixin abstract right? 

Why is this needed?

File: source/gui/contextHelp.py (hunk 9 above)
Comment (feerrenrut) [reply source/gui/contextHelp.py:79]: I'm sorry, I don't follow. I have already added the mixin to the settingsPanel base class.

File: source/gui/contextHelp.py (hunk 8 above)
Comment (leonardder) [reply source/gui/contextHelp.py:76]: Yes, I'm sorry. I forgot to mention that this depends on the assumption that helpId is an abstract property as I suggested.

File: source/gui/contextHelp.py (hunk 9 above)
Comment (leonardder) [reply source/gui/contextHelp.py:83]: Ah, I of course should have been clearer.

### Current situation, mixin added to SettingsPanel
* Every SettingsPanel has the mixin, so is expected to provide context help
* This doesn't apply to add-ons that use SettingsPanel, and there are plenty of them nowadays.

### SUggested situation
* Remove mixin from SettingsPanel
* Add mixin to every SettingsPanel NVDA core offers. Could be done using a SettingsPanelWithContextHelp intermediate class that inherrits from SettingsPanel and the mixin, though I don't care about being a bit more explicit here
* Then, mark helpId as an abstract property. It ensures that every panel or dialog that doesn't have helpId simply fails to initialize.

```
	@abstractproperty
	def helpId(self):
		raise NotImplementedError

```

Comment (feerrenrut) [reply source/gui/contextHelp.py:83]: I'd prefer not to do that. I think it's fair that every settings panel should have help, and for users to be informed that the addon panels do not have context help available.

Comment (leonardder) [reply source/gui/contextHelp.py:83]: Currently, users are only informed with a warning in the log. I'd agree if the help event would be bound regardless and show a message box indicating that there's no help.
Otherwise, adding a mixing to something that does nothing with it honestly really feels a bit strange to me.

Comment (feerrenrut) [reply source/gui/contextHelp.py:83]: > Currently, users are only informed with a warning in the log

Ah yes, this is a mistake, note that there is a user message when the helpId is empty.

**Comment by feerrenrut** [comment]
I believe all review comments from @leonardder have been addressed. It would be good to get this merged.

**Review by michaelDCurran — DISMISSED** [review]

**Comment by feerrenrut** [comment]
@michaelDCurran I think you approved this, but since the review was dismissed there is no record. Can you confirm?

**Comment by feerrenrut** [comment]
@michaelDCurran Never mind, I found the email notification. It was approved, I'll go ahead and merge this.
